subsys/net/lib/http/http_app.c

```diff
@@ -0,0 +1,264 @@
+	}
+
+	http_send_flush(ctx, NULL);
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+#if defined(CONFIG_HTTP_SERVER) && defined(CONFIG_NET_DEBUG_HTTP_CONN)
+	if (!ctx->is_client) {
+		http_server_conn_del(ctx);
+	}
+#endif
+
+	return net_app_close(&ctx->app_ctx);
+}
+
+int http_release(struct http_ctx *ctx)
+{
+	if (!ctx) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -ENOENT;
+	}
+
+	ctx->is_tls = false;
+
+#if defined(CONFIG_HTTP_SERVER) && defined(CONFIG_NET_DEBUG_HTTP_CONN)
+	if (!ctx->is_client) {
+		http_server_conn_del(ctx);
+		http_server_disable(ctx);
+	}
+#endif
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	ctx->is_init = false;
+
+	return net_app_release(&ctx->app_ctx);
+}
+
+int http_send_msg_raw(struct http_ctx *ctx, struct net_pkt *pkt,
+		      void *user_send_data)
+{
+	int ret;
+
+	NET_DBG("[%p] Sending %zd bytes data", ctx, net_pkt_get_len(pkt));
+
+	ret = net_app_send_pkt(&ctx->app_ctx, pkt, NULL, 0, 0,
+			       user_send_data);
+	if (!ret) {
+		/* We must let the system to send the packet, otherwise TCP
+		 * might timeout before the packet is actually sent. This is
+		 * easily seen if the application calls this functions many
+		 * times in a row.
+		 */
+		k_yield();
+	}
+
+	return ret;
+}
+
+static int _http_add_chunk(struct http_ctx *ctx, const u8_t *buf, size_t len,
+			   void *user_send_data)
+{
+	char chunk_header[16];
+	int ret;
+
+	if (!buf) {
+		len = 0;
+	}
+
+	snprintk(chunk_header, sizeof(chunk_header), "%x" HTTP_CRLF,
+		 (unsigned int)len);
+
+	ret = http_prepare_and_send(ctx, chunk_header, strlen(chunk_header),
+				    user_send_data);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (len) {
+		ret = http_prepare_and_send(ctx, buf, len, user_send_data);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = http_prepare_and_send(ctx, HTTP_CRLF, sizeof(HTTP_CRLF),
+				    user_send_data);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+int http_prepare_and_send(struct http_ctx *ctx,
+			  const char *payload,
+			  size_t payload_len,
+			  void *user_send_data)
+{
+	size_t added;
+	int ret;
+
+	do {
+		if (!ctx->pending) {
+			ctx->pending = net_app_get_net_pkt(&ctx->app_ctx,
+							   AF_UNSPEC,
+							   ctx->timeout);
+			if (!ctx->pending) {
+				return -ENOMEM;
+			}
+		}
+
+		ret = net_pkt_append(ctx->pending, payload_len, payload,
+				     ctx->timeout);
+
+		added = ret;
+
+		if (added < payload_len) {
+			/* Not all data could be added, send what we have now
+			 * and allocate new stuff to be sent.
+			 */
+			ret = http_send_flush(ctx, user_send_data);
+			if (ret < 0) {
+				return ret;
+			}
+
+			payload_len -= added;
+			payload += added;
+		}
+	} while (added < payload_len);
+
+	return 0;
+}
+
+int http_send_chunk(struct http_ctx *ctx,
+		    const char *payload,
+		    size_t payload_len,
+		    void *user_send_data)
+{
+	return _http_add_chunk(ctx, payload, payload_len, user_send_data);
```

Comment:
Do we need _http_add_chunk() at all, can you put its body here?

Reply:
Yep, that one is unnecessary func.